### PR TITLE
Experiment with possible strategies for exception handling

### PR DESCRIPTION
**Pull request by Nachasic**
In this PR I am experimenting with possible ways to handle platform API exceptions in convenient manner, while converting them to custom strongly-typed errors for ease of use. Directly related to https://github.com/lensapp/lenscloud-lens-extension/issues/119

### Basics - `/exceptions/common.exceptions.ts`
This file contains three main things:
1. `interface LensPlatformExceptionData` to describe the shape of data we want to put in our custom exceptions;
1. `LensPlatformException` - base class for custom exceptions, extends `Error`, implements `LensPlatformExceptionData` and stores `rawException` in a public reference for debugging purposes;
1. `NotFoundException` an example of a custom exception class, inherited from `LensPlatformException`;

### Utilities for painless handling in SDK functions - `/exceptions/utils.ts`
This file contains `expectExceptions()` function. in arguments it accepts a closure and an `HTTPErrorCode => LensPlatformException` mapping (typed as `HTTPErrCodeExceptionMap`). This function would execute a closure in a `try-catch` block. If closure is executed successfully, it returns the result of the closure. If it catches errors, it will then try to extract an HTTP error code from a caught exception, look it up in the mapping provided. If the map contains the extracted code, it then will throw an appropriate custom exception. If the code is not found in the mapping, it will produce a basic `LensSDKException` with an HTTP error code it caught, an appropriate text message and raw error stored within custom exception. If it fails to extract HTTP error code from caught error, it will throw a `400 (Bad Request)` custom exception. 

### Exemplary usage - `Users.ts`
I rewrote `UserService.getOne()` method to illustrate how all this could be utilized:
```ts
import { expectExceptions, NotFoundException } from "./exceptions";

class UserService extends Base {
  async getOne({ username }: { username: string }): Promise<User> {
    const { apiEndpointAddress, got } = this.lensPlatformClient;
    const url = `${apiEndpointAddress}/users/${username}`;

    const json = expectExceptions(
      () => got.get(url),
      { 404: () => new NotFoundException(`User ${username} not found`) }
    );

    return (json as unknown) as User;
  }
  // ...
}
```
## What's the point?
1. Make exception handling more explicit and type-safe;
2. Provide SDK users with a way to distinguish between different exceptions for implementation of robust and safe handling strategies.

### CODE REVIEW

**Comment by Nachasic** [comment]
Todo:
1. [ ] Make sure the whole thing works as designed, provide basic tests;
1. [ ] Refactor `expectExceptions()` so it won't throw unless asked explicitly;
1. [ ] Refactor global error handler to work nicely with `expectExceptions()` (right now it won't, because global error handling happens on a deeper stack layers inside `got()`);
1. [ ] Create custom exceptions for relevant cases for `lens-platform-extension`, test interoperability;
1. [ ]  Set up E2E tests;
1. [ ] Create custom exceptions for every function in SDK, where it's relevant.

**Comment by Nachasic** [comment]
Starting to think having a global exception handler is not such a good idea, at least in a way it's currently implemented. Take `UserService.getOne()` for example:
```typescript
class UserService extends Base {
  async getOne({ username }: { username: string }): Promise<User> {
    const { apiEndpointAddress, got } = this.lensPlatformClient;
    const url = `${apiEndpointAddress}/users/${username}`;

    const json = throwExpected(
      () => got.get(url),
      { 404: () => new NotFoundException(`User ${username} not found`) }
    );

    return (json as unknown) as User;
  }
```
Just by looking at this function one can deduce several things:
1. It returns a `User` as a promise;
2. If it returns successfully, the shape of the resulting data is always going to be consistent with the `User` type;
3. It throws exceptions when data cannot be properly obtained.

When we introduce global a exception handler that operates on deeper levels of stack (e.g. in `got` as it is currently implemented), this effectively changes the behavior of this method: now any exception thrown from `got.get()` might be handled, which would mean that instead of exiting prematurely, the method would continue as per execution order and assign an `undefined` value to `const json`, which then gets returned. As a result we get a function that we think would either produce a valid `User` or throw, yet now it returns an `undefined` value which is not reflected neither in type definitions, nor in the definition of this function itself.

To put it simply, global exception handler seems magical to me, and not in a good way.

**Comment by Nachasic** [comment]
### Update
I've provided exemplary exception handling for `UserService`. Things to note:
1. It is possible to provide all custom exception constructors  with context-aware messages:
    ```ts
    const foo = new NotFoundException(); // has default error message
    const bar = new NotFoundException("Could not find cluster in a space"); // has custom error message
    ```
1. `HTTPErrCodeExceptionMap` now allows exception creators to accept original error as an optional argument for more nuanced handling. I've implemented this upon realization that one API function can sometimes throw multiple different errors with the same HTTP error code, e.g. `SpaceService.getClusterToken()`. Examples of implementation coming soon.

File: src/Space.ts [hunk 1]
```diff
@@ -54,7 +69,11 @@ class SpaceService extends Base {
   async getMany(queryString?: string): Promise<Space[]> {
     const { apiEndpointAddress, got } = this.lensPlatformClient;
     const url = `${apiEndpointAddress}/spaces${queryString ? `/?${queryString}` : ""}`;
-    const json = await got.get(url);
+
+    const json = await throwExpected(
+      () => got.get(url),
+      {}
```
Comment (panuhorsmalahti) [inline src/Space.ts:75]: We could make this parameter optional, defaulting to `{}`.

In that case `throwExpected` seems unnecessary? (But maybe ok for consistency).

Comment (Nachasic) [reply src/Space.ts:75]: Optional mapping is a great idea.
I've included `throwExpected` everywhere because it also handles unexpected exceptions by converting them to custom `LensSDKException` with debugging info.

File: src/Space.ts [hunk 2]
```diff
@@ -80,9 +102,16 @@ class SpaceService extends Base {
     const { apiEndpointAddress, got } = this.lensPlatformClient;
     const url = `${apiEndpointAddress}/spaces/${spaceName}`;
 
-    const json = await got.patch(url, {
-      json: space
-    });
+    const json = await throwExpected(
+      () => got.patch(url, {
```
Comment (panuhorsmalahti) [inline src/Space.ts:106]: What if we use switch and case for this?

```
try {
    const json = got.patch(url, { json: space });
catch (error) {
    // assuming .status_code here
    switch (error.status_code) {
        case 500:
        throw new TokenNotFoundException();

        // ...

       default:
       throw error;
    }
}
```

Comment (Nachasic) [reply src/Space.ts:105]: That's the problem: while debugging API errors in the extension I noticed that status code is often not easily accessible - that's why I had to write a separate function that attempts to extract it. Unless I'm missing something.

The mapping could definitely be refactored to simple closure with switch-case logic, but that would result in a requirement to provide default implementation for unexpected errors - which right now is done automatically in `throwExpected()`

Comment (panuhorsmalahti) [reply src/Space.ts:105]: Makes sense

File: src/Exceptions.test.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  beforeAll(() => {
+    nock(apiEndpointAddress).get(`/users/${invalidUserName}`).reply(404, plaform404response);
+    nock(apiEndpointAddress).get(`/users/${brokenUserName}`).reply(500, plaform500response).persist();
+    nock(apiEndpointAddress).get("/users/test").reply(200, { });
+  });
+
+  it("doesn't break functionality", async () => {
+    const result = await lensPlatformClient.user.getOne({ username: "test" });
+
+    expect(result).toStrictEqual({});
```
Comment (chenhunghan) [inline src/Exceptions.test.ts:30]: [nit] Minor refactoring to make the test more robust
```suggestion
  const username = "a_random_user"
  const getOneResponse = { username };
  beforeAll(() => {
    nock(apiEndpointAddress).get(`/users/${invalidUserName}`).reply(404, plaform404response);
    nock(apiEndpointAddress).get(`/users/${brokenUserName}`).reply(500, plaform500response).persist();
    nock(apiEndpointAddress).get(`/users/${username}`).reply(200, getOneResponse);
  });

  it("doesn't break functionality", async () => {
    const result = await lensPlatformClient.user.getOne({ username });

    expect(result).toStrictEqual(getOneResponse);
```

File: src/LensPlatformClient.test.ts [hunk 4]
```diff
@@ -75,14 +71,17 @@ describe("LensPlatformClient", () => {
...
+      try {
+        await Promise.all([
+          _got.get(apiEndpointAddress),
+          _got.post(apiEndpointAddress),
+          _got.put(apiEndpointAddress),
+          _got.patch(apiEndpointAddress),
+          _got.head(apiEndpointAddress),
+          _got.delete(apiEndpointAddress)
+        ]);
+      } catch (e: unknown) {}
```
Comment (chenhunghan) [inline src/LensPlatformClient.test.ts:84]: Maybe put the "spies tests" in the finally block? Or add a comment to explain why we want to catch the error and ignore it here. (non trivial for me when reading here) 

```suggestion
      } finally {
        spies.forEach(spy => {
          // tests are moved here
        }
      }
```

File: src/LensPlatformClient.ts [hunk 5]
```diff
@@ -182,11 +182,13 @@ class LensPlatformClient {
...
-              } else {
-                throw error;
-              }
+              throw error;
+              // Disable exception handler for testing purposes
+              // if (exceptionHandler) {
+              //   exceptionHandler(error);
+              // } else {
+              //   throw error;
+              // }
```
Comment (chenhunghan) [inline src/LensPlatformClient.ts:191]: I think you can remove the commented (I fully agree `exceptionHandler` is bad by design.
```suggestion
```

File: src/Exceptions.test.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    } catch (e: unknown) {
+      // Check exact exception
+      expect(e).toStrictEqual(expectedException);
+      // Check reflected type
+      expect(e).toBeInstanceOf(NotFoundException);
+      expect(e).toBeInstanceOf(LensSDKException);
+    }
+  });
+
+  it("thows proper unexpected exception", async () => {
```
Comment (chenhunghan) [inline src/Exceptions.test.ts:47]: typo
```suggestion
  it("throws proper unexpected exception", async () => {
```

File: src/Exceptions.test.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    try {
+      const _ = await lensPlatformClient.user.getOne({ username: brokenUserName });
+    } catch (e: unknown) {
+      expect(e).toBeInstanceOf(LensSDKException);
+      expect((e as LensSDKException).errorCode).toEqual(500);
+      expect((e as LensSDKException).message).toMatch(/Unexpected exception \[Lens Platform SDK\]/);
```
Comment (chenhunghan) [inline src/Exceptions.test.ts:53]: Maybe this instead of type casting?
```suggestion
      if (e instanceof LensSDKException) {
        expect((e).errorCode).toEqual(500);
        expect((e).message).toMatch(/Unexpected exception \[Lens Platform SDK\]/);
      }
```

File: src/Exceptions.test.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    const result = await lensPlatformClient.user.getOne({ username: "test" });
+
+    expect(result).toStrictEqual({});
+  });
+
+  it("throws proper expected exception", async () => {
+    const expectedException = new NotFoundException(`User ${invalidUserName} not found`);
+
+    try {
+      const _ = await lensPlatformClient.user.getOne({ username: invalidUserName });
```
Comment (chenhunghan) [inline src/Exceptions.test.ts:37]: we probably don't need `_`
```suggestion
      await lensPlatformClient.user.getOne({ username: invalidUserName });
```

File: src/Exceptions.test.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+      expect(e).toStrictEqual(expectedException);
+      // Check reflected type
+      expect(e).toBeInstanceOf(NotFoundException);
+      expect(e).toBeInstanceOf(LensSDKException);
+    }
+  });
+
+  it("thows proper unexpected exception", async () => {
+    try {
+      const _ = await lensPlatformClient.user.getOne({ username: brokenUserName });
```
Comment (chenhunghan) [inline src/Exceptions.test.ts:49]: we probably don't need `_`
```suggestion
     await lensPlatformClient.user.getOne({ username: brokenUserName });
```

**Comment by Nachasic** [comment]
While discussing this with @panuhorsmalahti and @chenhunghan we've decided that this direction of exception handling is worth pursuing. It has been pointed out that there are use cases for global exception handling mechanism (e.g. for handling "token expired" exceptions that could be thrown basically anywhere in the SDK), but this should be the responsibility of code that incorporates the SDK. There is also need for e2e tests - @panuhorsmalahti agreed to help with that.

**Comment by chenhunghan** [comment]
@Nachasic Please ping us when it's ready for review

File: integration-test/SpaceService.test.ts [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+  UnauthorizedException,
+  SpaceNameReservedException,
+  SpaceNotFoundException,
+  ForbiddenException
+} from "../src/exceptions";
+
+const TEST_SPACE_NAME = "test-space";
+const [credBob, credAlice] = config.users;
+
+describe.only("SpaceService", () => {
```
Comment (chenhunghan) [inline integration-test/SpaceService.test.ts:15]:
```suggestion
describe("SpaceService", () => {
```

File: .github/workflows/test.yaml [hunk 11]
```diff
@@ -18,5 +27,10 @@ jobs:
     - uses: actions/setup-node@v2
       with:
         node-version: ${{ matrix.node-version }}
+    # there is a problem in node 16 in CI
+    # which can't find eslint/eslint from dependencies
```
Comment (chenhunghan) [inline .github/workflows/test.yaml:31]: Sorry, my typo
```suggestion
    # which can't find eslint/jest from dependencies
```

File: integration-test/SpaceService.test.ts [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    const description = "My space description";
+    const updatedDescription = "New description";
+
+    let space = await testPlatformBob.client.space.createOne({ name, description });
+    expect(space.name).toEqual(name);
+
+    space = await testPlatformBob.client.space.updateOne(name, { name, description: updatedDescription });
+    expect(space.description).toEqual(updatedDescription);
+
+    await testPlatformBob.client.space.deleteOne({ name: name });
```
Comment (chenhunghan) [inline integration-test/SpaceService.test.ts:40]: Maybe this to ensure the space get deleted even if tests are not pass.
```suggestion
    try {
      let space = await testPlatformBob.client.space.createOne({ name, description });
      expect(space.name).toEqual(name);

      space = await testPlatformBob.client.space.updateOne(name, { name, description: updatedDescription });
      expect(space.description).toEqual(updatedDescription);
    } finally {
      await testPlatformBob.client.space.deleteOne({ name: name });
    }
```

File: integration-test/README.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
+# Integration tests
+
+The integration tests are executed against the staging environment. The GitHub Action secrets are specified to provide the required environmental variables.
+
+The integration tests expect `STAGING_USERNAME` and `STAGING_PASSWORD` environmental variables when running the tests:
```
Comment (panuhorsmalahti) [inline integration-test/README.md:5]: These should be updated (STAGING_USERNAME_1 etc)

File: integration-test/UserService.test.ts [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        .rejects.toThrowError(UnauthorizedException);
+    });
+
+    it("can get users", async () => {
+      const users = await bobPlatform.client.user.getMany(`filter=email||$eq||${userBob.username}@mirantis.com`);
+      expect(users.length).toEqual(0);
+    });
+
+    it("rejects bad requests", async () =>
+      expect(
```
Comment (chenhunghan) [inline integration-test/UserService.test.ts:83]: To you need to return this expect like in other tests?
```suggestion
      return expect(
```

File: integration-test/SpaceService.test.ts [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    testPlatformBob = await testPlatformFactory(credBob.username, credBob.password);
+  });
+
+  beforeEach(() => {
+    testPlatformBob.fakeToken = undefined;
+    testPlatformAlice.fakeToken = undefined;
+  });
+
+  it("allows to create, update and delete a space", async () => {
+    const name = `create-${rng()}`;
```
Comment (panuhorsmalahti) [inline integration-test/SpaceService.test.ts:30]: We should prefix all the names with something clearly indicating they belong to integration tests, such as `integration-test-${rng()}`

File: integration-test/UserService.test.ts (hunk 14 above)
Comment (chenhunghan) [reply integration-test/UserService.test.ts:83]: (just wonder if `bobPlatform.client.user.getMany()` will be a floating promise) but I am not sure how `rejects.toThrowError()` works

**Review by chenhunghan — APPROVED** [review]
Overall, great work!, thanks for taking this library to the next level, I add some inline comments, there are not critical but please consider the suggested changes, I also tag the PR with `major`, so the next release to main will be the `major` version.

File: integration-test/UserService.test.ts [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+        bobPlatform.client.user.getMany()
+      ).rejects.toThrowError(BadRequestException)
+    );
+  });
+
+  describe("getSelf", () => {
+    it("rejects requiests with invalid tokens", async () => {
```
Comment (panuhorsmalahti) [inline integration-test/UserService.test.ts:90]: typo: requiests

File: src/SpaceService.ts [hunk 17]
```diff
@@ -106,7 +143,15 @@ class SpaceService extends Base {
     const { apiEndpointAddress, got } = this.lensPlatformClient;
     const url = `${apiEndpointAddress}/spaces/${name}/k8sclusters/${clusterId}/bored-secret`;
 
-    const json = await got.get(url);
+    const json = await throwExpected(
+      () => got.get(url),
+      {
+        // eslint-disable-next-line
```
Comment (panuhorsmalahti) [inline src/SpaceService.ts:149]: Why is `// eslint-disable-next-line` needed?

File: src/SpaceService.ts [hunk 18]
```diff
@@ -169,7 +237,15 @@ class SpaceService extends Base {
...
     const url = `${apiEndpointAddress}/spaces/${cluster.space?.name}/k8sclusters/${cluster.id}`;
 
-    const json = await got.put(url, { json: cluster });
+    const json = await throwExpected(
+      () => got.put(url, { json: cluster }),
+      {
+        // eslint-disable-next-line
+        // TODO: differentiate between space and cluster not being found
+        404: () => new SpaceNotFoundException(cluster.space?.name ?? "undefined"),
+        400: () => new BadRequestException("Property 'kind' of cluster object is invalid")
```
Comment (panuhorsmalahti) [inline src/SpaceService.ts:246]: I think there are multiple reasons for the bad request, not just wrong kind.

File: src/SpaceService.ts [hunk 19]
```diff
@@ -201,7 +295,9 @@ class SpaceService extends Base {
     const { apiEndpointAddress, got } = this.lensPlatformClient;
     const url = `${apiEndpointAddress}/spaces/${name}/plan`;
 
-    const json = await got.get(url);
+    const json = await throwExpected(
+      () => got.get(url)
```
Comment (panuhorsmalahti) [inline src/SpaceService.ts:299]: This can return not found

File: src/UserService.ts [hunk 20]
```diff
@@ -62,7 +70,13 @@ class UserService extends Base {
   async updateOne(username: string, user: User & { attributes?: UserAttributes } & { password?: string }): Promise<User> {
     const { apiEndpointAddress, got } = this.lensPlatformClient;
     const url = `${apiEndpointAddress}/users/${username}`;
-    const json = await got.patch(url, { json: user });
+    const json = await throwExpected(
+      () => got.patch(url, { json: user }),
+      {
```
Comment (panuhorsmalahti) [inline src/UserService.ts:75]: Bad request exception missing

File: src/exceptions/HTTPErrrorCodes.ts [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+  502: "Bad Gateway",
+  503: "Service Unavailable",
+  504: "Gateway Timeout",
+  505: "HTTP Version Not Supported",
+  506: "Variant Also Negotiates",
+  507: "Insufficient Storage",
+  508: "Loop Detected",
+  510: "Not Extended",
+  511: "Network Authentication Required"
+};
```
Comment (panuhorsmalahti) [inline src/exceptions/HTTPErrrorCodes.ts:41]: Could we improve this by adding `as const` at the end?

File: src/exceptions/invitation.exceptions.ts [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+export class UserAlreadyExistsException extends UnprocessableEntityException {
+  constructor(userName?: string) {
+    super(`User${userName ? " @" + userName : ""} already exists in space`);
+    Object.setPrototypeOf(this, UserAlreadyExistsException.prototype);
+  }
+}
+
+export class PendingInvitationException extends UnprocessableEntityException {
+  constructor(userName?: string) {
+    super(`User${userName ? " @" + userName : ""} already has a pending invitation to this space`);
```
Comment (panuhorsmalahti) [inline src/exceptions/invitation.exceptions.ts:19]: Why not `User ${userName ? " " + userName : ""} already has`...

File: src/exceptions/invitation.exceptions.ts [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    Object.setPrototypeOf(this, PendingInvitationException.prototype);
+  }
+}
+
+export class EmailMissingException extends UnprocessableEntityException {
+  constructor() {
+    super("E-mail address is missing from invitation object");
```
Comment (panuhorsmalahti) [inline src/exceptions/invitation.exceptions.ts:26]: I'd use "Email" instead

File: src/exceptions/user.exceptions.ts [hunk 24]
```diff
@@ -0,0 +1,15 @@
+import { NotFoundException, LensSDKException } from "./common.exceptions";
+
+export class UserNameNotFoundException extends NotFoundException {
+  constructor(userName: string) {
+    super(`Could not find user @${userName}`);
```
Comment (panuhorsmalahti) [inline src/exceptions/user.exceptions.ts:5]: Maybe remove "@"

**Comment by panuhorsmalahti** [comment]
Good work 👍

File: integration-test/UserService.test.ts (hunk 14 above)
Comment (Nachasic) [reply integration-test/UserService.test.ts:83]: I'm not sure as well, I think returning it is a better idea

Comment (Nachasic) [reply integration-test/UserService.test.ts:83]: On the other hand, `return` keyword is not needed here, because the whole test is packed into an arrow closure - it returns automatically

File: src/SpaceService.ts (hunk 18 above)
Comment (Nachasic) [reply src/SpaceService.ts:242]: Certainly, there are a bunch of places where existing error handling needs improvements and testing. I think it would make sense to do this in subsequent PRs.

Comment (Nachasic) [reply src/SpaceService.ts:242]: Yes, cluster functions still need some work. I was thinking of delegating this part to subsequent PRs

File: integration-test/UserService.test.ts (hunk 14 above)
Comment (chenhunghan) [reply integration-test/UserService.test.ts:83]: ah, you are right

File: src/SpaceService.ts (hunk 19 above)
Comment (Nachasic) [reply src/SpaceService.ts:294]: Same as above, will do in later PRs

File: src/UserService.ts (hunk 20 above)
Comment (Nachasic) [reply src/UserService.ts:75]: Same as above - later PRs
